org.abapgit.adt.ui/src/org/abapgit/adt/ui/internal/wizards/AbapGitWizardPageObjectsSelectionForPull.java

```diff
@@ -222,81 +195,190 @@ public Object[] getElements(Object inputElement) {
 				return false;
 			}
 
 			@Override
 			public boolean isChecked(Object element) {
+				if (element instanceof IRepositoryModifiedObjects) {
+					IRepositoryModifiedObjects repo = (IRepositoryModifiedObjects) element;
+					Set<String> selectedNames = AbapGitWizardPageObjectsSelectionForPull.this.selectedObjectNamesByRepo
+							.get(repo.getRepositoryURL());
+					if (selectedNames == null) {
+						return false;
+					}
+					return selectedNames.size() == repo.getModifiedObjects().size();
+				}
 				if (element instanceof IOverwriteObject) {
-					IOverwriteObject object = (IOverwriteObject) element;
-					if (object.getAction() != null && object.getAction().equals("1")) { //$NON-NLS-1$
-						return true;
+					IOverwriteObject obj = (IOverwriteObject) element;
+					IRepositoryModifiedObjects parent = findParent(obj);
+					if (parent == null) {
+						return false;
 					}
+					Set<String> selectedNames = AbapGitWizardPageObjectsSelectionForPull.this.selectedObjectNamesByRepo
+							.get(parent.getRepositoryURL());
+					return selectedNames != null && selectedNames.contains(obj.getName());
 				}
 				return false;
 			}
+
+			private IRepositoryModifiedObjects findParent(IOverwriteObject obj) {
+				for (IRepositoryModifiedObjects repo : AbapGitWizardPageObjectsSelectionForPull.this.repoToModifiedObjects) {
+					if (repo.getModifiedObjects().contains(obj)) {
+						return repo;
+					}
+				}
+				return null;
+			}
 		});
+
+		// Scroll-based lazy loading
+		var tree = this.modifiedObjTreeViewer.getTree();
+		tree.addListener(SWT.MouseWheel, e -> tryLazyLoad());
+		tree.addListener(SWT.Selection, e -> tryLazyLoad());
+		tree.addListener(SWT.Resize, e -> tryLazyLoad());
+		tree.addListener(SWT.MouseUp, e -> tryLazyLoad());
 	}
-}
 
+	private void tryLazyLoad() {
+		var tree = this.modifiedObjTreeViewer.getTree();
+		ScrollBar vBar = tree.getVerticalBar();
+		if (vBar == null || vBar.getMaximum() == 0) {
+			return;
+		}
+
+		int scrollPosition = vBar.getSelection() + vBar.getThumb();
+		// Trigger loading when scrolled to 90% of the maximum for smoother experience
+		int threshold = (int) (vBar.getMaximum() * 0.9);
+
+		if (scrollPosition >= threshold) {
+			Object[] expandedElements = this.modifiedObjTreeViewer.getExpandedElements();
+
+			for (Object parent : expandedElements) {
+				if (parent instanceof IRepositoryModifiedObjects) {
+					IRepositoryModifiedObjects repo = (IRepositoryModifiedObjects) parent;
+					loadNextBatch(repo);
+				}
+			}
+		}
+	}
+
+	private void loadNextBatch(IRepositoryModifiedObjects repo) {
+		List<IOverwriteObject> all = repo.getModifiedObjects();
+		int currentOffset = this.repoOffsets.getOrDefault(repo, 0);
+
+		if (currentOffset >= all.size()) {
+			return;
+		}
+
+		int nextOffset = Math.min(currentOffset + ModifiedObjectTreeContentProvider.BATCH_SIZE, all.size());
+		List<IOverwriteObject> nextBatch = all.subList(currentOffset, nextOffset);
+		this.repoOffsets.put(repo, nextOffset);
```

Comment:
The offSet concept is a bit tricky when it comes to readability. 
My suggestion would be to rename variable currentOffset to itemsAlreadyLoaded.
Then calculate the number of items to load next as follows:

```
int itemsToLoadNext = Math.min(ModifiedObjectTreeContentProvider.BATCH_SIZE, 
                                     all.size() - itemsAlreadyLoaded); 
                                     
List<IOverwriteObject> nextBatch = allModifiedObjects.subList(itemsAlreadyLoaded, 
                                                                  itemsAlreadyLoaded + itemsToLoadNext);                 

this.repoOffsets.put(repo, itemsToLoadNext + itemsAlreadyLoaded );
                                                                                      
```